experimental/tools/cpp-simulation/cpp-simulation.cpp

```diff
@@ -0,0 +1,2237 @@
+
+    outsData = insData[dataCast<APInt>(index->data).getZExtValue()];
+  }
+
+  void printStates() override {
+    for (unsigned i = 0; i < size; ++i)
+      printValue<ConsumerRW, const Data>("ins", ins[i], insData[i].data);
+    printValue<ProducerRW, Data>("outs", outs, outsData.data);
+    printValue<ChannelProducerRW, Data>("index", index, &index->data);
+  }
+
+private:
+  handshake::ControlMergeOp op;
+
+  // parameters
+  unsigned size, indexWidth;
+
+  // internal components
+  TEHBSupport cMergeTEHB;
+  ForkSupport cMergeFork;
+
+  // ports
+  std::vector<ConsumerRW *> ins;
+  ProducerRW *outs;
+  ChannelProducerRW *index;
+
+  std::vector<ConsumerData> insData;
+  ProducerData outsData;
+
+  // control merge dataless
+  Data indexTEHB = APInt(indexWidth, 0);
+  bool dataAvailable = false, readyToFork = false, tehbOutValid = false,
+       tehbOutReady = false;
+  ConsumerRW insTEHB;
+  ProducerRW outsTEHB;
+  ConsumerRW insFork;
+  std::vector<ProducerRW *> outsFork;
+};
+
+class ForkModel : public OpExecutionModel<handshake::ForkOp> {
+public:
+  using OpExecutionModel<handshake::ForkOp>::OpExecutionModel;
+  ForkModel(handshake::ForkOp forkOp, mlir::DenseMap<Value, RW *> &subset)
+      : OpExecutionModel<handshake::ForkOp>(forkOp), op(forkOp),
+        size(op->getNumResults()), forkSupport(size),
+        ins(getState<ConsumerRW>(op.getOperand(), subset)), insData(ins) {
+    for (unsigned i = 0; i < size; ++i)
+      outs.push_back(getState<ProducerRW>(op->getResult(i), subset));
+
+    for (unsigned i = 0; i < size; ++i)
+      outsData.emplace_back(outs[i]);
+
+    forkSupport.datawidth = insData.dataWidth;
+  }
+
+  void reset() override { forkSupport.reset(ins, outs, insData, outsData); }
+
+  void exec(bool isClkRisingEdge) override {
+    forkSupport.exec(isClkRisingEdge, ins, outs, insData, outsData);
+  }
+
+  void printStates() override {
+    printValue<ConsumerRW, const Data>("ins", ins, insData.data);
+    for (unsigned i = 0; i < size; ++i)
+      printValue<ProducerRW, Data>("outs", outs[i], outsData[i].data);
+  }
+
+private:
+  handshake::ForkOp op;
+
+  // parameters
+  unsigned size;
+  ForkSupport forkSupport;
+
+  // ports
+  ConsumerRW *ins;
+  std::vector<ProducerRW *> outs;
+
+  ConsumerData insData;
+  std::vector<ProducerData> outsData;
+};
+
+// Never used but let it be
+class JoinModel : public OpExecutionModel<handshake::JoinOp> {
+public:
+  using OpExecutionModel<handshake::JoinOp>::OpExecutionModel;
+  JoinModel(handshake::JoinOp joinOp, mlir::DenseMap<Value, RW *> &subset)
+      : OpExecutionModel<handshake::JoinOp>(joinOp), op(joinOp),
+        join(op->getNumOperands()) {
+    for (auto oper : op->getOperands())
+      ins.push_back(getState<ConsumerRW>(oper, subset));
+    outs = getState<ProducerRW>(op.getResult(), subset);
+  }
+
+  void reset() override { join.exec(ins, outs); }
+
+  void exec(bool isClkRisingEdge) override { reset(); }
```

Comment:
Didn't realize this was an implication of our design for asynchronous components which also don't care about `reset`. In that case `exec` is the same as `reset`, which I find relatively elegant!